README
pyqtdeploy - A Tool for Deploying PyQt4 and PyQt5 Applications
==============================================================

What is pyqtdeploy?
-------------------

pyqtdeploy is a tool that, in conjunction with other tools provided with Qt,
enables the deployment of PyQt applications written with either Python v2.7 or
Python v3.3 or later.  It supports deployment to desktop platforms (Linux,
Windows and macOS) and to mobile platforms (iOS and Android).

pyqtdeploy itself requires PyQt5 and Python v3.3 or later.

pyqtdeploy works by taking the individual modules of a PyQt application,
freezing them, and then placing them in a Qt resource file that is converted to
C++ code by Qt's ``rcc`` tool.  Python's standard library is handled in the
same way.

pyqtdeploy generates a simple C++ wrapper around the Python interpreter library
that uses the Python import mechanism to enable access to the embedded frozen
modules in a similar way that Python supports the packaging of modules in zip
files.

Finally pyqtdeploy generates a Qt ``.pro`` file that describes all the
generated C++ code.  From this Qt's ``qmake`` tool is used to generate a
platform-specific ``Makefile`` which will then generate a single executable.
Further Qt and/or platform specific tools can then be used to convert the
executable to a platform specific deployable package.

When run pyqtdeploy presents a GUI that allows the modules that make up the
application to be specified, along with the PyQt components that the
application requires and the parts of the Python standard library that should
also be included.  This information is stored in a pyqtdeploy project file.
pyqtdeploy can also be run as a command line tool to generate the C++ code from
a project file.


Author
------

pyqtdeploy is copyright (c) Riverbank Computing Limited.  Its homepage is
https://www.riverbankcomputing.com/software/pyqtdeploy/.

Support may be obtained from the PyQt mailing list at
https://www.riverbankcomputing.com/mailman/listinfo/pyqt


License
-------

pyqtdeploy is released under the BSD license.


Installation
------------

pyqtdeploy can be dowloaded an installed from PyPI::

    pip3 install pyqtdeploy

pyqtdeploy requires PyQt5 to be installed.  This is not installed
automatically.


Documentation
-------------

The documentation for the latest release can be found
`here <http://pyqt.sourceforge.net/Docs/pyqtdeploy/>`__.


NEWS
v2.4.1 ??? ??? 2019
  - Added support for Python v3.7.3.

v2.4 22nd March 2019
  - Added android-64 as a target architecture.
  - Added support for Python v2.7.16, v3.4.10, v3.5.7, v3.6.8 and v3.7.2.
  - Added support for PyQt5.QtRemoteObjects.
  - Added support for PyQtWebEngine as an add-on package.
  - Added support for Android NDK r19.
  - Added 'android_ndk_version', 'android_sdk_version',
    'android_toolchain_cc', 'android_toolchain_cflags' and 'verify_source()' to
    the sysroot API.

v2.3.1 2nd December 2018
  - Added the 'edition' option to the Qt5 component plugin.

v2.3 17th November 2018
  - Added support for Python v3.6.7 and v3.7.1.
  - Added the 'version' option to the Python component plugin.
  - The 'source' option of the Python component plugin is now only required if
    building either the host or target versions from source.

v2.2 30th August 2018
  - Added support for Python v2.7.15, v3.4.8, v3.4.9, v3.5.5, v3.5.6, v3.6.5,
    v3.6.6 and v3.7.0.
  - Added support for PyQt v5.11.3 and later and private sip modules.  (Earlier
    versions of PyQt v5.11 are not supported.)
  - Added support for the 'importlib.resoures' feature added in Python v3.7.
  - Added the zlib component plugin.
  - Added the 'no_asm' option to the openssl component plugin.
  - Added support for OpenSSL v1.1 to the openssl component plugin.
  - Added the 'host_installation_bin_dir' option to the python component
    plugin.
  - A directory specified by the '--source-dir' option of pyqtdeploy-sysroot
    may now be read-only.
  - The '--source-dir' option of pyqtdeploy-sysroot may now be specified any
    number of times.
  - The demo now displays the source code of the demo in a separate tab.
  - The '--source-dir' option of build-demo.py may now be specified any number
    of times.
  - The '--installed-qt-dir' option of build-demo.py has been added.
  - Added 'add_to_path()', 'android_ndk_sysroot', 'android_toolchain_bin' and
    'android_toolchain_prefix' to the sysroot API.

v2.1 30th January 2018
  - Added support for PyQt5.QtNetworkAuth.
  - Added support for PyQt5.Qt3DAnimation.

v2.0.1 5th January 2018
  - A bug fix only release.

v2.0 5th January 2018
  - Added pyqtdeploy-build.
  - Added pyqtdeploy-sysroot.
  - Removed pyqtdeploycli.
  - Added support for Python v2.7.14, v3.4.7, v3.5.4, v3.6.1, v3.6.2, v3.6.3
    and v3.6.4.
  - Added support for PyQt5.QtAndroidExtras.
  - Added the pyqt-demo demonstration application.

v1.3.2 4th March 2017
  - Added support for Python v2.7.13, v3.4.6, v3.5.3 and v3.6.0.
  - Added support for PyQt3D.
  - Added support for PyQt5.QtWebEngine.
  - The default Android API level is now 21 (i.e. Android v5.0).
  - Added the --timeout command line option to pyqtdeploycli.
  - The experimental build-sysroot.py script will now build OpenSSL on Windows
    and macOS.

v1.3.1 30th August 2016
  - Added support for Python v2.7.12, v3.4.5 and v3.5.2.
  - Added support for PyQt v5.7.
  - Added support for Qt v5.7.0.

v1.3 14th June 2016
  - The Build tab can now specify a timeout for build processes to override the
    default of 30 seconds.
  - The Standard Library tab now has a checkbox to specify that sub-packages
    should be automatically selected.
  - Added the --android-api command line option to pyqtdeploycli to specify the
    Android API level to use when configuring the Python package.  The default
    API level is extracted from the ANDROID_NDK_PLATFORM environment variable.
    If that is not set then 9 is used.
  - Added support for PyQt v5.6.
  - Added support for Qt v5.6.0.
  - The experimental build-sysroot.py script has been re-written and
    documented.

v1.2.1 13th March 2016
  - A bug fix only release.

v1.2 8th January 2016
  - Added the install action and --sysroot and --system-python command line
    options to pyqtdeploycli to copy Python from a standard Windows
    installation.
  - Added the --disable-patches command line option to pyqtdeploycli to prevent
    the patching of the Python source code.
  - Added the --include-dir, --interpreter, --python-library, --source-dir and
    --standard-library-dir command line options to pyqtdeploycli to override
    values specified in the project file.
  - Added the $PDY_PY_MAJOR, $PDY_PY_MINOR and $PDY_PY_MICRO pseudo-environment
    variables that can be embedded in file and directory paths specified in the
    Locations tab and are automatically expanded to the appropriate values.
  - The importer now supports the PEP 302 get_data(), get_code(), get_source()
    and is_package() optional import hooks.
  - Added support for Python v2.7.11, v3.4.4, v3.5.0 and v3.5.1.
  - Added support for the bdb and pdb modules.

v1.1 18th August 2015
  - The Other Extensions tab can now be used to compile simple extension
    modules (including those generated by Cython) as well as link in
    pre-compiled modules.
  - Added support for PyQt v5.5.
  - Added support for Python v2.7.10.

v1.0.1 6th May 2015
  - Added support for the pathlib module.

v1.0 2nd March 2015
  - Added support for Python v3.4.3.

v0.10 4th February 2015
  - All GUI elements now have What's This? help.
  - Added the show-version action to pyqtdeploycli.

v0.9 28th January 2015
  - Added support for Python v2.7.9.
  - Added support for PyQt v5.4.
  - Additional qmake configuration statements can be added.
  - Applications can now be built against shared system installations of
    Python.
  - The number of resource files generated can now be increased in order to
    work around problems with MSVC on Windows.

v0.8 15th October 2014
  - A bug fix only release.

v0.7 14th October 2014
  - Added support for Python v3.4.2 and v3.3.6
  - Updated the documentation for building a static version of Qt on Windows to
    include statically linking the C library.

v0.6 13th September 2014
  - Support for Python v2.6 has been removed.
  - Split into two separate applications: the GUI pyqtdeploy and the command
    line pyqtdeploycli.

pyqtdeploy/main_cli.py
# Copyright (c) 2016, Riverbank Computing Limited
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
# 
# 1. Redistributions of source code must retain the above copyright notice,
#    this list of conditions and the following disclaimer.
# 
# 2. Redistributions in binary form must reproduce the above copyright notice,
#    this list of conditions and the following disclaimer in the documentation
#    and/or other materials provided with the distribution.
# 
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.


import argparse
import os
import sys


def main():
    """ The entry point for the setuptools generated CLI wrapper. """

    # Get the default Android API level.  This is the level that Python v3.6.0
    # targets.
    default_api = 21

    parts = os.environ.get('ANDROID_NDK_PLATFORM', '').split('-')

    if len(parts) == 2 and parts[0] == 'android':
        try:
            default_api = int(parts[1])
        except ValueError:
            pass

    # Parse the command line.
    parser = argparse.ArgumentParser()

    parser.add_argument('action',
            help="the action to perform",
            choices=('build', 'configure', 'install', 'show-packages',
                    'show-targets', 'show-version'))
    parser.add_argument('--android-api',
            help="the Android API level to target when configuring Python "
                    "(configure) [default: {}]".format(default_api),
            metavar="LEVEL", type=int, default=default_api),
    parser.add_argument('--disable-patches',
            help="disable the patching of the Python source code (configure)",
            action='store_true')
    parser.add_argument('--enable-dynamic-loading',
            help="enable the dynamic loading of modules (configure)",
            action='store_true')
    parser.add_argument('--include-dir',
            help="the target Python include directory (build)", metavar="DIR")
    parser.add_argument('--interpreter',
            help="the host interpreter executable (build)",
            metavar="EXECUTABLE")
    parser.add_argument('--opt',
            help="the optimisation level where 0 is none, 1 is no asserts, 2 "
                    "is no asserts or docstrings (build) [default: 2]",
            metavar="LEVEL", type=int, choices=range(3), default=2),
    parser.add_argument('--output',
            help="the name of the output file or directory (configure, build)",
            metavar="OUTPUT")
    parser.add_argument('--package',
            help="the package name (configure, install)",
            metavar="PACKAGE")
    parser.add_argument('--project', help="the project file (build)",
            metavar="FILE")
    parser.add_argument('--python-library',
            help="the target Python library (build)", metavar="LIB")
    parser.add_argument('--resources',
            help="the number of .qrc resource files to generate (build) "
                    "[default: 1]",
            metavar="NUMBER", type=int, default=1),
    parser.add_argument('--source-dir',
            help="the Python source code directory (build)", metavar="DIR")
    parser.add_argument('--standard-library-dir',
            help="the target Python standard library directory (build)",
            metavar="DIR")
    parser.add_argument('--sysroot',
            help="the system image root directory (install)",
            metavar="DIR")
    parser.add_argument('--system-python',
            help="use the system installed Python (install)",
            metavar="VERSION")
    parser.add_argument('--target',
            help="the target platform (configure, install)",
            metavar="TARGET")
    parser.add_argument('--timeout',
            help="the number of seconds to wait for build processes to run "
                    "before timing out (build)",
            metavar="SECONDS", type=int, default=30)
    parser.add_argument('--quiet', help="disable progress messages (build)",
            action='store_true')
    parser.add_argument('--verbose',
            help="enable verbose progress messages (configure, build)",
            action='store_true')

    args = parser.parse_args()

    # Handle the specific actions.
    if args.action == 'build':
        rc = build(args)
    elif args.action == 'configure':
        rc = configure(args)
    elif args.action == 'install':
        rc = install(args)
    elif args.action == 'show-packages':
        rc = show_packages(args)
    elif args.action == 'show-targets':
        rc = show_targets(args)
    elif args.action == 'show-version':
        rc = show_version(args)
    else:
        # This should never happen.
        rc = 1

    return rc


def build(args):
    """ Perform the build action. """

    if args.project is None:
        missing_argument('--project')
        return 2

    if args.resources < 1:
        print(
                "{0}: error: argument --resources: number must be at least 1".format(
                        os.path.basename(sys.argv[0])),
                file=sys.stderr)
        return 2

    from . import Builder, MessageHandler, Project, UserException

    message_handler = MessageHandler(args.quiet, args.verbose)

    try:
        builder = Builder(Project.load(args.project), message_handler)
        builder.build(args.opt, args.resources, args.timeout,
                build_dir=args.output, include_dir=args.include_dir,
                interpreter=args.interpreter,
                python_library=args.python_library, source_dir=args.source_dir,
                standard_library_dir=args.standard_library_dir)
    except UserException as e:
        handle_exception(e, args.verbose)
        return 1

    return 0


def configure(args):
    """ Perform the configure action. """

    if args.package is None:
        missing_argument('--package')
        return 2

    if args.package == 'python':
        from . import configure_python, MessageHandler, UserException

        message_handler = MessageHandler(args.quiet, args.verbose)

        try:
            configure_python(args.target, args.output, args.android_api,
                    args.enable_dynamic_loading, not args.disable_patches,
                    message_handler)
        except UserException as e:
            handle_exception(e, args.verbose)
            return 1
    else:
        from . import configure_package, UserException

        try:
            configure_package(args.package, args.target, args.output)
        except UserException as e:
            handle_exception(e, args.verbose)
            return 1

    return 0


def install(args):
    """ Perform the install action. """

    # Note that the intent is to support the installation of all supported
    # packages into sysroot in a future version.  For now we just handle the
    # Windows version of Python installed from the official installers.

    if args.package is None:
        missing_argument('--package')
        return 2

    if args.sysroot is None:
        missing_argument('--sysroot')
        return 2

    if args.package == 'python':
        from . import install_python, MessageHandler, UserException

        message_handler = MessageHandler(args.quiet, args.verbose)

        try:
            install_python(args.target, args.sysroot, args.system_python,
                    message_handler)
        except UserException as e:
            handle_exception(e, args.verbose)
            return 1
    else:
        error("install only supports the python package at the moment.")
        return 1

    return 0


def show_packages(args):
    """ Perform the show-packages action. """

    from . import get_supported_packages, UserException

    try:
        packages = get_supported_packages()
    except UserException as e:
        handle_exception(e, args.verbose)
        return 1

    show(packages)

    return 0


def show_targets(args):
    """ Perform the show-targets action. """

    from . import get_supported_targets, UserException

    try:
        targets = get_supported_targets()
    except UserException as e:
        handle_exception(e, args.verbose)
        return 1

    show(targets)

    return 0


def show_version(args):
    """ Perform the show-version action. """

    from . import PYQTDEPLOY_RELEASE

    print(PYQTDEPLOY_RELEASE)


def show(items):
    """ Show an unsorted list of items on stdout. """

    for item in sorted(items):
        print(item)


def missing_argument(name):
    """ Tell the user about a missing argument. """

    # Mimic the argparse message.
    error("error: the following arguments are required: {0}".format(name))


def handle_exception(e, verbose):
    """ Tell the user about an exception. """

    if verbose and e.detail != '':
        error("{0}: {1}".format(e.text, e.detail))
    else:
        error(e.text)


def error(message):
    """ Tell the user about an error. """

    print("{0}: {1}".format(os.path.basename(sys.argv[0]), message),
            file=sys.stderr)
